intelmq/bin/intelmqctl.py

```diff
@@ -346,7 +349,237 @@ def __status_process(self, pid, module, bot_id):
+                                  "You may want to launch: 'intelmqctl stop {}' to prevent this message."
+                                  .format(bot_id))
+            paused = True
+            self.bot_stop(bot_id)
+
+        log_bot_message("starting", bot_id)
+
+        try:
+            BotDebugger(self.__runtime_configuration[bot_id], bot_id, run_subcommand,
+                        console_type, message_action_kind, dryrun, msg, show_sent,
+                        loglevel=loglevel)
+            retval = 0
+        except KeyboardInterrupt:
+            print("Keyboard interrupt.")
+            retval = 0
+        except SystemExit as exc:
+            print("Bot exited with code %s." % exc.code)
+            retval = exc.code
+
+        if paused:
+            self.bot_start(bot_id)
+
+        return retval
+
+    def bot_start(self, bot_id: str, getstatus=True):
+        state = self._get_process_state(bot_id)
+        if state is not None:
+            if state == self.ProcessState.RUNNING:
+                log_bot_message("running", bot_id)
+                return "running"
+
+            elif state not in (self.ProcessState.STARTING, self.ProcessState.BACKOFF):
+                self._remove_bot(bot_id)
+
+        log_bot_message("starting", bot_id)
+        self._create_and_start_bot(bot_id)
+
+        if getstatus:
+            return self.bot_status(bot_id)
+
+    def bot_stop(self, bot_id: str, getstatus=True):
+        state = self._get_process_state(bot_id)
+        if state is None:
+            if not self.__controller._is_enabled(bot_id):
+                log_bot_message("disabled", bot_id)
+                return "disabled"
+            else:
+                log_bot_error("stopped", bot_id)
+                return "stopped"
+
+        if state not in (self.ProcessState.RUNNING, self.ProcessState.STARTING, self.ProcessState.BACKOFF):
+            self._remove_bot(bot_id)
+            log_bot_error("stopped", bot_id)
+            return "stopped"
+
+        log_bot_message("stopping", bot_id)
+
+        self._get_supervisor().supervisor.stopProcess(self._process_name(bot_id))
+        self._remove_bot(bot_id)
+
+        if getstatus:
+            return self.bot_status(bot_id)
+
+    def bot_reload(self, bot_id: str, getstatus=True):
+        state = self._get_process_state(bot_id)
+        if state is None:
+            if not self.__controller._is_enabled(bot_id):
+                log_bot_message("disabled", bot_id)
+                return "disabled"
+            else:
+                log_bot_error("stopped", bot_id)
+                return "stopped"
+
+        if state not in (self.ProcessState.RUNNING, self.ProcessState.STARTING, self.ProcessState.BACKOFF):
+            self._remove_bot(bot_id)
+            log_bot_error("stopped", bot_id)
+            return "stopped"
+
+        log_bot_message("reloading", bot_id)
+        self._get_supervisor().supervisor.signalProcess(self._process_name(bot_id), "HUP")
```

Comment:
Reloading did not work:
```
$ intelmqctl reload abusech-feodo-domains-collector
intelmqctl: Connected to supervisor 'supervisor' (API version 3.0)
intelmqctl: Reloading bot abusech-feodo-domains-collector ...
Traceback (most recent call last):
  File "/usr/local/bin/intelmqctl", line 11, in <module>
    load_entry_point('intelmq', 'console_scripts', 'intelmqctl')()
  File "/home/sebastianw/intelmq/intelmq/bin/intelmqctl.py", line 1377, in main
    return x.run()
  File "/home/sebastianw/intelmq/intelmq/bin/intelmqctl.py", line 869, in run
    retval, results = args.func(**args_dict)
  File "/home/sebastianw/intelmq/intelmq/bin/intelmqctl.py", line 902, in bot_reload
    status = self.bot_process_manager.bot_reload(bot_id, getstatus)
  File "/home/sebastianw/intelmq/intelmq/bin/intelmqctl.py", line 478, in bot_reload
    self._get_supervisor().supervisor.signalProcess(self._process_name(bot_id), "HUP")
  File "/usr/lib/python3.4/xmlrpc/client.py", line 1089, in __call__
    return self.__send(self.__name, args)
  File "/usr/lib/python3.4/xmlrpc/client.py", line 1419, in __request
    verbose=self.__verbose
  File "/usr/lib/python3.4/xmlrpc/client.py", line 1131, in request
    return self.single_request(host, handler, request_body, verbose)
  File "/usr/lib/python3.4/xmlrpc/client.py", line 1147, in single_request
    return self.parse_response(resp)
  File "/usr/lib/python3.4/xmlrpc/client.py", line 1318, in parse_response
    return u.close()
  File "/usr/lib/python3.4/xmlrpc/client.py", line 656, in close
    raise Fault(**self._stack[0])
xmlrpc.client.Fault: <Fault 1: 'UNKNOWN_METHOD'>
```

That has been added in 3.2.0 but I have 3.0.0 (Debian jessie): http://supervisord.org/changes.html#id11
Maybe we can catch this?

Reply:
I added better error message and also warning when old supervisor version is used.